operator/v1alpha1/types_externaldns.go

```diff
@@ -0,0 +1,366 @@
+}
+
+// +kubebuilder:validation:Enum=OpenShiftRoute;Service;CRD
+type ExternalDNSSourceType string
+
+const (
+	SourceTypeRoute   ExternalDNSSourceType = "OpenShiftRoute"
+	SourceTypeService ExternalDNSSourceType = "Service"
+	SourceTypeCRD     ExternalDNSSourceType = "CRD"
+)
+
+// +kubebuilder:validation:Enum=ignore;allow
+type HostnameAnnotationPolicy string
+
+const (
+	HostnameAnnotationPolicyIgnore HostnameAnnotationPolicy = "Ignore"
+	HostnameAnnotationPolicyAllow  HostnameAnnotationPolicy = "Allow"
+)
+
+// ExternalDNSServiceSourceOptions describes options
+// specific to the externalDNS service source.
+type ExternalDNSServiceSourceOptions struct {
+	// PublishPolicy determines what kinds of Service
+	// resources are published to the given DNS provider.
+	//
+	// The following publishing policies are available:
+	//
+	//  "PublishExternal": Only publish DNS records for
+	//  service resources that externally reachable,
+	//  such as NodePort, ExternalName, and LoadBalancer
+	//  services. This is the default behavior of externalDNS.
+	//
+	//  "PublishInternalExternal": Publish DNS records
+	//   for externally reachable services, in addition to
+	//   internally reachable services, such as ClusterIP
+	//   services.
+	//
+	// +optional
+	PublishPolicy *ServicePublishPolicy `json:"publishPolicy,omitempty"`
+
+	// ServiceType determines what type of Service resources
+	// are watched by externalDNS. The following types are
+	// available options:
+	//
+	//  "NodePort"
+	//  "ExternalName"
+	//  "LoadBalancer"
+	//  "ClusterIP" (requires a publishPolicy of "PublishInternalExternal")
+	//
+	// +optional
+	ServiceType *corev1.ServiceType `json:"serviceType,omitempty"`
+}
+
+type ServicePublishPolicy string
+
+// +kubebuilder:validation:Enum=PublishExternal;PublishInternalExternal
+const (
+	ServicePublishPolicyExternalOnly        ServicePublishPolicy = "PublishExternal"
+	ServicePublishPolicyExternalAndInternal ServicePublishPolicy = "PublishInternalExternal"
+)
+
+type ExternalDNSCRDSourceOptions struct {
+	// Resource is the kind of the CRD
+	// source resource type to be
+	// consumed by externalDNS.
+	//
+	// e.g. "DNSEndpoint"
+	//
+	// +optional
+	Resource *string `json:"resource,omitempty"`
+
+	// Version is the API version
+	// of the given resource kind for
+	// externalDNS to use.
+	//
+	// e.g. "externaldns.k8s.io/v1alpha1"
+	//
+	// +optional
+	Version *string `json:"version,omitempty"`
+
+	// LabelFilter specifies a label filter
+	// to be used to filter CRD resource instances.
+	// Only one label filter can be specified on
+	// an externalDNS instance.
+	//
+	// +optional
+	LabelFilter *string `json:"labelFilter,omitempty"`
+}
+
+// ExternalDNSStatus defines the observed state of externalDNS
+type ExternalDNSStatus struct {
+	// Conditions is a list of operator-specific conditions
+	// and their status.
+	Conditions []ExternalDNSOperatorCondition `json:"conditions,omitempty"`
```

Comment:
I wanted to avoid coupling this API to the base Operator API in case we upstream this API. Hence why I copied and renamed `OperatorCondition`. Does that sound like a reasonable thing to do for the time being, or should I just fall back to using `OperatorCondition`?

Reply:
That's a fair point.  However, we'll need to change things when we upstream the API anyway (if nothing else, the API group will change to use the upstream namespace), and `ExternalDNSOperatorCondition` and `OperatorCondition` are basically the same, aren't they?  I don't feel strongly about this in any case.

Reply:
Ah, now I remember why I created the new `ExternalDNSOperatorCondition` type.
`OperatorCondition` doesn't user points for optional values, but the `ExternalDNSOperatorCondition` type does.

Reply:
Let's leave `ExternalDNSOperatorCondition` for now and see if API reviewers have any input (or if they decide we can't use pointers for optional fields).